### PR TITLE
NAS-101349 / 11.3 / Use Alerts for UPS events

### PR DESCRIPTION
**Pull request by sonicaj**
This PR introduces changes where we don't spam terminals on each UPS events and instead use Alerts system to notify user of different UPS events.
Ticket: #NAS-101349

### CODE REVIEW

File: src/middlewared/middlewared/alert/source/ups.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    category = AlertCategory.UPS
+    level = AlertLevel.INFO
+    title = 'UPS Communications established'
+    text = 'Communications with UPS %(ups)s established.'
+
+
+class UPSReplbattAlertClass(AlertClass, SimpleOneShotAlertClass):
+    category = AlertCategory.UPS
+    level = AlertLevel.CRITICAL
+    title = 'UPS Battery needs to be replaced'
```
Comment (william-gr) [inline src/middlewared/middlewared/alert/source/ups.py]: We should be using title case for every word
USB Battery Needs To Be Replace

Thats the unspoken rule we have been using, iirc

Comment (wblock) [reply src/middlewared/middlewared/alert/source/ups.py]: Yes, these should be title case.  But title case is not every word, there are odd rules.
There are online things that will do it correctly, mostly.

File: src/middlewared/middlewared/alert/source/ups.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+from middlewared.alert.base import SimpleOneShotAlertClass, AlertClass, AlertCategory, AlertLevel
+
+
+class UPSBatteryLowAlertClass(AlertClass, SimpleOneShotAlertClass):
+    category = AlertCategory.UPS
+    level = AlertLevel.ALERT
+    title = 'UPS Battery LOW'
+    text = 'UPS %(ups)s Battery level low.'
```
Comment (wblock) [inline src/middlewared/middlewared/alert/source/ups.py]:
```suggestion
    text = 'UPS %(ups)s battery level low.'
```

File: src/middlewared/middlewared/alert/source/ups.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+
+class UPSOnlineAlertClass(AlertClass, SimpleOneShotAlertClass):
+    category = AlertCategory.UPS
+    level = AlertLevel.INFO
+    title = 'UPS ONLINE'
```
Comment (wblock) [inline src/middlewared/middlewared/alert/source/ups.py]:
```suggestion
    title = 'UPS On Line Power'
```

File: src/middlewared/middlewared/alert/source/ups.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    level = AlertLevel.CRITICAL
+    title = 'UPS On Battery'
+    text = 'UPS %(ups)s is on battery power.'
+
+
+class UPSCommbadAlertClass(AlertClass, SimpleOneShotAlertClass):
+    category = AlertCategory.UPS
+    level = AlertLevel.CRITICAL
+    title = 'UPS Communication Lost'
+    text = 'Communications with UPS %(ups)s lost.'
```
Comment (wblock) [inline src/middlewared/middlewared/alert/source/ups.py]:
```suggestion
    text = 'Communication with UPS %(ups)s lost.'
```

File: src/middlewared/middlewared/alert/source/ups.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+
+class UPSCommokAlertClass(AlertClass, SimpleOneShotAlertClass):
+    category = AlertCategory.UPS
+    level = AlertLevel.INFO
+    title = 'UPS Communications Established'
```
Comment (wblock) [inline src/middlewared/middlewared/alert/source/ups.py]:
```suggestion
    title = 'UPS Communication Established'
```

File: src/middlewared/middlewared/alert/source/ups.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    text = 'Communications with UPS %(ups)s established.'
```
Comment (wblock) [inline src/middlewared/middlewared/alert/source/ups.py]:
```suggestion
    text = 'Communication with UPS %(ups)s established.'
```

File: src/middlewared/middlewared/alert/source/ups.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+
+class UPSReplbattAlertClass(AlertClass, SimpleOneShotAlertClass):
+    category = AlertCategory.UPS
+    level = AlertLevel.CRITICAL
+    title = 'UPS Battery Needs To Be Replaced'
```
Comment (wblock) [inline src/middlewared/middlewared/alert/source/ups.py]:
```suggestion
    title = 'UPS Battery Needs Replacement'
```

File: src/middlewared/middlewared/alert/source/ups.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    text = 'UPS %(ups)s Battery needs to be replaced.'
```
Comment (wblock) [inline src/middlewared/middlewared/alert/source/ups.py]:
```suggestion
    text = 'UPS %(ups)s battery needs replacement.'
```

File: src/middlewared/middlewared/plugins/ups.py [hunk 9]
```diff
@@ -191,10 +191,31 @@ def driver_choices(self):
         old_config['toemail'] = ';'.join(old_config['toemail']) if old_config['toemail'] else ''
 
         if len(set(old_config.items()) ^ set(config.items())) > 0:
+            if config['identifier'] != old_config['identifier']:
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/ups.py]: But you are going to restart UPS service anyway and they will be dismissed there.

File: src/middlewared/middlewared/plugins/ups.py [hunk 10]
```diff
@@ -191,10 +191,31 @@ def driver_choices(self):
...
+            'ONBATT': 'UPSOnBattery',
+            'ONLINE': 'UPSOnline',
+            'REPLBATT': 'UPSReplbatt'
+        }
+
+    @private
+    async def dismiss_alerts(self):
+        config = await self.config()
+
+        for alert in (await self.alerts_mapping).values():
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/ups.py]: `await self.alerts_mapping()`

File: src/middlewared/middlewared/plugins/ups.py [hunk 11]
```diff
@@ -191,10 +191,31 @@ def driver_choices(self):
...
+            await self.middleware.call('alert.oneshot_delete', alert, {'ups': config['identifier']})
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/ups.py]: We need to dismiss them for all UPSes.
Probably it's not possible with current alerts implementation and you might need to extend it a little.

File: src/middlewared/middlewared/plugins/ups.py (hunk 9 above)
Comment (sonicaj) [reply src/middlewared/middlewared/plugins/ups.py]: I am not so sure about that - if the identifier changes - new identifier will be used to dismiss them and that wouldn't work ?

**Pull request merged**